### PR TITLE
add DeployGate gradle plugin

### LINKED ISSUE #202: [Assigned] Use DeployGate gradle plugin
## Overview (Required)
currently we are using a simple curl command, but it would be better to use this official plugin.
https://github.com/DroidKaigi/conference-app-2018/blob/69d8b4282558822281a9a128f20902b88af5219d/deploy.sh

https://docs.deploygate.com/docs/gradle-plugin

### PR DESCRIPTION
**Pull request by kitwtnb**
## Issue
- close #202 

## Overview (Required)
- Add DeployGate gradle plugin

Should I need to delete `deploy.sh` ? 🤔 

## Links
- https://docs.deploygate.com/docs/gradle-plugin

### CODE REVIEW

**Comment by konifar** [comment]
Thanks! 

> Should I need to delete deploy.sh ?

Yes, can you change the apk upload logic for CI too?

**Comment by kitwtnb** [comment]
Sorry, I don't know alot about the upload logic for CI tool. 😢

**Comment by konifar** [comment]
Got it. It's okay.

Current upload script is this.
https://github.com/DroidKaigi/conference-app-2018/blob/master/deploy.sh

Deploygate upload CI task is here.
https://github.com/DroidKaigi/conference-app-2018/blob/master/.circleci/config.yml#L70

Can you replace this command to gradle command which is provided by DeployGate gradle plugin?

```
./gradlew :app:uploadDeployGateDebug
```

**Comment by kitwtnb** [comment]
Thank you!! 😭 
I fixed it, this ok?

**Comment by konifar** [comment]
You have to setup the gradle for DeployGate upload.
https://docs.deploygate.com/docs/gradle-plugin-configuration

Can you try?

**Comment by mhidaka** [comment]
39e01d0 quick feedback! 🆒

File: build.gradle [hunk 1]
```diff
@@ -16,6 +16,7 @@ buildscript {
         classpath "gradle.plugin.org.jlleitschuh.gradle:ktlint-gradle:3.0.0"
         classpath 'io.fabric.tools:gradle:1.25.1'
         classpath 'com.github.ben-manes:gradle-versions-plugin:0.17.0'
+        classpath 'com.deploygate:gradle:1.1.3'
```
Comment (kwmt) [inline build.gradle]: I know the version number is written in [DeployGate document](https://docs.deploygate.com/docs/gradle-plugin#section-1-build-gradle-), but latest version is 1.1.4. You can check it on [bintray.com](https://bintray.com/deploygate/maven/gradle).

**Comment by kitwtnb** [comment]
I tried fix referring to `deploy.sh`.
What do you think?

```java
deploygate {
    token = System.getenv()['DEPLOY_GATE_API_KEY']

    apks {
        debug {
            def hash = 'git rev-parse --short HEAD'.execute([], project.rootDir).in.text.trim()
            message = "https://github.com/DroidKaigi/conference-app-2018/tree/${hash} https://circleci.com/gh/DroidKaigi/conference-app-2018/${System.getenv()['CIRCLE_BUILD_NUM']}"

            distributionKey = "aed2445665e27de6571227992d66ea489b6bdb44"
            releaseNote = "https://github.com/DroidKaigi/conference-app-2018/tree/${hash} https://circleci.com/gh/DroidKaigi/conference-app-2018/${System.getenv()['CIRCLE_BUILD_NUM']}"
        }
    }
}
```

**Comment by konifar** [comment]
Almost good!
- Please add username to run on CI.
- You can get Env variable like this `System.getenv("DEPLOY_GATE_API_KEY")`

```
deploygate {
    userName = "takahirom"
    token = System.getenv("DEPLOY_GATE_API_KEY")

    apks {
        release {
            def hash = 'git rev-parse --short HEAD'.execute([], project.rootDir).in.text.trim()
            message = "https://github.com/DroidKaigi/conference-app-2018/tree/${hash} https://circleci.com/gh/DroidKaigi/conference-app-2018/${System.getenv("CIRCLE_BUILD_NUM")}"

            distributionKey = "aed2445665e27de6571227992d66ea489b6bdb44"
            releaseNote = "https://github.com/DroidKaigi/conference-app-2018/tree/${hash} https://circleci.com/gh/DroidKaigi/conference-app-2018/${System.getenv("CIRCLE_BUILD_NUM")}"
        }
    }
}
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -67,7 +67,7 @@ jobs:
               fi
       - run:
           name: Deploy
-          command: ./deploy.sh
+          command: ./gradlew :app:uploadDeployGateDebug
```
Comment (konifar) [inline .circleci/config.yml]: Maybe `./gradlew uploadDeployGateRelease` is correct.
Plus, this command should be run when the code is merged to only master branch.
https://github.com/DroidKaigi/conference-app-2018/blob/master/deploy.sh#L3

Comment (konifar) [reply .circleci/config.yml]: Can you add the branch check logic like this?
https://github.com/DroidKaigi/conference-app-2018/pull/204/files#diff-1d37e48f9ceff6d8030570cd36286a61R63

**Comment by kitwtnb** [comment]
I corrected it, but I failed the build. 😢 
Is the option setting incorrect?

File: .circleci/config.yml [hunk 3]
```diff
@@ -67,7 +67,10 @@ jobs:
               fi
       - run:
           name: Deploy
-          command: ./deploy.sh
+          command: |
+              if [ ! $CIRCLE_BRANCH = 'master' ]; then
```
Comment (konifar) [inline .circleci/config.yml]: This is incorrect.
Deploygate upload task should be run when the circle branch is `master` branch.

Comment (kitwtnb) [reply .circleci/config.yml]: I'm sorry to trouble you. 😭 
I succeed to build, thanks to you!

**Comment by konifar** [comment]
Thanks for fixing! and sorry! can you resolve the conflict? 🙇

**Comment by kitwtnb** [comment]
I corrected conflict, please confirm this!

File: build.gradle [hunk 4]
```diff
@@ -14,6 +14,7 @@ buildscript {
         classpath "gradle.plugin.org.jlleitschuh.gradle:ktlint-gradle:$versions.ktlintGradle"
         classpath "io.fabric.tools:gradle:$versions.fabricGradleTool"
         classpath "com.github.ben-manes:gradle-versions-plugin:$versions.gradleVersions"
+        classpath "com.deploygate:gradle:$versions.deploygate"
```
Comment (konifar) [inline build.gradle]: 👍

**Comment by konifar** [comment]
Thanks for fixing! It seems perfect for me.
Let's merge and check this will work on master branch build 😉 
If it will now work well, don't worry. we can fix soon. Actually Deploygate member is in DroidKaigi staff.

Anyway, thanks for your contribution! LGTM!

**Comment by kitwtnb** [comment]
I have learned to CI tool setting, thanks to you. :joy:
:tada:

**Pull request merged**